File: src/lib/ebay-review-inbox.ts
import type { PriceReviewQueue } from "./admin-price-review";

export function mergeEbayReviewInbox(queue: PriceReviewQueue, inbox: PriceReviewQueue): PriceReviewQueue {
  const existing = new Set(queue.items.map((item) => item.id));
  const decided = new Set(queue.decisions.map((decision) => String(decision.id ?? "")));
  const reviewed = new Map(inbox.items.filter((item) => {
    const decision = item.decision;
    return item.source.startsWith("ebay") && decision?.reviewBatch && decision.reviewer
      && (decision.action === "accept" || decision.action === "reject")
      && item.status === (decision.action === "accept" ? "accepted" : "rejected")
      && inbox.decisions.some((entry) => entry.id === item.id && entry.reviewBatch === decision.reviewBatch
        && entry.action === decision.action && entry.catalogId === decision.catalogId);
  }).map((item) => [item.id, item]));
  const imported = new Set<string>();
  const items = queue.items.map((item) => {
    const replacement = reviewed.get(item.id);
    if (item.status === "pending" && !decided.has(item.id) && replacement) {
      imported.add(item.id);
      return replacement;
    }
    return item;
  });
  const additions = inbox.items.filter((item) => {
    if (!item.source.startsWith("ebay") || existing.has(item.id) || decided.has(item.id)) return false;
    if (item.status !== "pending" && !reviewed.has(item.id)) return false;
    if (reviewed.has(item.id)) imported.add(item.id);
    existing.add(item.id);
    return true;
  });
  return { ...queue, items: [...items, ...additions], decisions: [
    ...queue.decisions, ...inbox.decisions.filter((entry) => imported.has(String(entry.id))),
  ] };
}

File: scripts/collectors/ebay_review_inbox.py
"""Additive eBay inbox consumed by the existing Admin review queue."""

from collectors.common import load_json, now_iso, save_json
from collectors.price_review_queue import REVIEW_KEYS, _row_to_item


def publish_review_inbox(path, ingest, platform, batch_id):
    previous = load_json(path, {"items": [], "decisions": []})
    items = {item["id"]: item for item in previous.get("items", [])}
    for key in REVIEW_KEYS:
        for row in ingest.get(key) or []:
            if not str(row.get("source", "")).startswith("ebay"):
                continue
            item = _row_to_item(row, row["source"], platform, ingest)
            if not item:
                continue
            old = items.get(item["id"])
            if old and old.get("status") in {"accepted", "rejected"}:
                continue
            item["jobId"] = batch_id
            item["createdAt"] = (old or {}).get("createdAt") or item["createdAt"]
            items[item["id"]] = item
    # Do not use the generic 5,000-item truncation: unrelated pending evidence
    # must not be silently removed as the eBay campaign grows.
    save_json(path, {"schemaVersion": 1, "updatedAt": now_iso(),
                     "items": list(items.values()), "decisions": previous.get("decisions", [])})
    return len(items)
